lib/avro/builder/definition_cache.rb

```diff
@@ -0,0 +1,64 @@
+module Avro
+  module Builder
+
+    # This class is used to cache previously defined schema objects
+    # preventing direct access via the DSL.
+    class DefinitionCache
+      def initialize(builder)
+        @builder = builder
+        @schema_objects = {}
+        @schema_names = Set.new
+      end
+
+      # Cache and schema object by name (for convenience) and fullname.
+      # The schema object is only available by name if the unqualified name is unique.
+      def add_schema_object(object)
+        store_by_name(object) if object.namespace
+        store_by_fullname(object)
+      end
+
+      # Lookup an Avro schema object by name, possibly fully qualified by namespace.
+      def lookup_named_type(key, namespace = nil)
+        key_str = Avro::Name.make_fullname(key.to_s, namespace && namespace.to_s)
+        object = schema_objects[key_str]
+
+        unless object || schema_names.include?(key.to_s)
```

Comment:
It might just be me but I always struggle to read `unless` conditions with compound predicates.